src/python/pants/engine/README.md

```diff
@@ -76,52 +76,80 @@ In cases where this search detects any ambiguity (generally because there are tw
 that can provide the same product with the same number of parameters), rule graph compilation will
 fail with a useful error message.
 
-### Datatypes
+### How to Declare Engine `Param` Types
```

Comment:
Consider something like `How to declare engine newtypes`. John and Tom Dyas used that phrase the other day and I think it would be great to leverage here because so many people already understand the concept of a newtype.

Reply:
Hm, ok. I'm pretty sure `newtype` is a specifically Haskell concept, and I don't know what it means? Could you be more specific about what concept is supposed to be more common knowledge here?

Reply:
Haskell has it, but so does Python with `typing.NewType` https://docs.python.org/3/library/typing.html#newtype. 

The concept is "I have some pre-existing type I like like `int`, but I want to create a new distinct type that acts like `int` but which must be explicitly used like `MyIntType`." 

The benefit of referencing `newtype` is that devs who already know that concept now instantly make the connection between their prior knowledge and Pants. Maybe calling it `new type` (with a space) + still explaining it, like this, would help people who still aren't familiar?

> The engine makes heavy use of the concept of _newtypes_ to disambiguate similar-looking types. For example, you might want to use `int` in several distinct domains, like `UserId` vs. `ProjectId`. New types allow you to avoid accidentally using `UserId` when you meant `ProjectId`, because each is now a distinct type.

Reply:
@cosmicexplorer The Rust ecosystem also uses the term "newtype" in explicit analogy to Haskell, even though there's not `newtype` keyword in Rust: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types

Reply:
I might add a line saying something like "These small, unique classes are sometimes called 'newtypes', since they effectively wrap an existing type in a 'new type' that the engine will treat as distinct from the underlying type.", maybe with a link to that Rust documentation page.